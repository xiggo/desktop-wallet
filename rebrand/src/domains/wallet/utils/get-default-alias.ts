import { Networks } from "@payvo/sdk";
import { Contracts } from "@payvo/sdk-profiles";

interface GetDefaultAliasInput {
	profile: Contracts.IProfile;
	network: Networks.Network;
}

export const getDefaultAlias = ({ profile, network }: GetDefaultAliasInput): string => {
	const makeAlias = (count: number) => `${network.displayName()} #${count}`;

	const sameCoinWallets = profile.wallets().findByCoinWithNetwork(network.coin(), network.id());

	let counter = sameCoinWallets.length;

	if (counter === 0) {
		counter = 1;
	}

	while (profile.wallets().findByAlias(makeAlias(counter))) {
		counter++;
	}

	return makeAlias(counter);
};
