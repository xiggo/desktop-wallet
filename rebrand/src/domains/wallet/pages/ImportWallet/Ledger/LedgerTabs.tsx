import { Enums } from "@payvo/sdk";
import { uniq } from "@payvo/sdk-helpers";
import React, { useCallback, useRef, useState } from "react";
import { useFormContext } from "react-hook-form";
import { useTranslation } from "react-i18next";
import { useHistory } from "react-router-dom";

import { LedgerConnectionStep } from "./LedgerConnectionStep";
import { LedgerImportStep } from "./LedgerImportStep";
import { LedgerScanStep } from "./LedgerScanStep";
import { LedgerTabsProperties, LedgerTabStep } from "./LedgerTabs.contracts";
import { Button } from "@/app/components/Button";
import { Icon } from "@/app/components/Icon";
import { StepIndicator } from "@/app/components/StepIndicator";
import { TabPanel, Tabs } from "@/app/components/Tabs";
import { LedgerData, useLedgerContext } from "@/app/contexts";
import { useActiveProfile } from "@/app/hooks";
import { useKeydown } from "@/app/hooks/use-keydown";
import { useWalletConfig } from "@/domains/wallet/hooks";
import { NetworkStep } from "@/domains/wallet/components/NetworkStep";
import { assertWallet } from "@/utils/assertions";

const Paginator = ({
	activeIndex,
	isMultiple,
	isNextDisabled,
	isNextLoading,
	onBack,
	onFinish,
	onNext,
	onRetry,
	showRetry,
	size,
}: {
	activeIndex: LedgerTabStep;
	isMultiple: boolean;
	isNextDisabled?: boolean;
	isNextLoading?: boolean;
	onBack: () => void;
	onFinish: () => void;
	onNext: () => void;
	onRetry?: () => void;
	showRetry?: boolean;
	size: number;
}) => {
	const { t } = useTranslation();

	return (
		<div className="flex justify-between mt-10">
			<div>
				{showRetry && (
					<Button variant="secondary" onClick={onRetry} data-testid="Paginator__retry-button">
						<Icon name="ArrowRotateRight" />
						<span>{t("COMMON.RETRY")}</span>
					</Button>
				)}
			</div>

			<div className="flex space-x-3">
				{activeIndex < size && (
					<Button variant="secondary" onClick={onBack} data-testid="Paginator__back-button">
						{t("COMMON.BACK")}
					</Button>
				)}

				{activeIndex < size && (
					<Button
						disabled={isNextDisabled || isNextLoading}
						isLoading={isNextLoading}
						onClick={onNext}
						data-testid="Paginator__continue-button"
					>
						{t("COMMON.CONTINUE")}
					</Button>
				)}

				{activeIndex === size && (
					<Button disabled={isNextDisabled} data-testid="Paginator__finish-button" onClick={onFinish}>
						{isMultiple ? t("COMMON.GO_TO_PORTFOLIO") : t("COMMON.GO_TO_WALLET")}
					</Button>
				)}
			</div>
		</div>
	);
};

export const LedgerTabs = ({
	activeIndex = LedgerTabStep.NetworkStep,
	onClickEditWalletName,
}: LedgerTabsProperties) => {
	const activeProfile = useActiveProfile();

	const history = useHistory();
	const { importLedgerWallets, isBusy } = useLedgerContext();
	const { selectedNetworkIds, setValue } = useWalletConfig({ profile: activeProfile });

	const { t } = useTranslation();

	const { formState, getValues, handleSubmit } = useFormContext();
	const { isValid, isSubmitting } = formState;

	const [importedWallets, setImportedWallets] = useState<LedgerData[]>([]);
	const [activeTab, setActiveTab] = useState<number>(activeIndex);

	const isMultiple = importedWallets.length > 1;

	const [showRetry, setShowRetry] = useState(false);
	const retryFunctionReference = useRef<() => void>();

	const importWallets = useCallback(
		async ({ network, wallets }: any) => {
			setImportedWallets(wallets);
			const coin = activeProfile.coins().set(network.coin(), network.id());
			await importLedgerWallets(wallets, coin, activeProfile);

			setValue("selectedNetworkIds", uniq([...selectedNetworkIds, coin.network().id()]));
		},
		[importLedgerWallets, activeProfile, setValue, selectedNetworkIds],
	);

	const isNextDisabled = isBusy || !isValid;

	useKeydown("Enter", () => {
		const isButton = (document.activeElement as any)?.type === "button";

		if (!isButton && !isNextDisabled && !isSubmitting) {
			if (activeTab < LedgerTabStep.LedgerImportStep) {
				handleNext();
			} else {
				handleFinish();
			}
		}
	});

	const handleNext = useCallback(async () => {
		if (activeTab === LedgerTabStep.LedgerScanStep) {
			await handleSubmit((data: any) => importWallets(data))();
		}

		setActiveTab(activeTab + 1);
	}, [activeTab, handleSubmit, importWallets]);

	const handleBack = () => {
		if (activeTab === LedgerTabStep.NetworkStep) {
			return history.push(`/profiles/${activeProfile.id()}/dashboard`);
		}

		if (activeTab === LedgerTabStep.LedgerScanStep) {
			return setActiveTab(LedgerTabStep.NetworkStep);
		}

		setActiveTab(activeTab - 1);
	};

	const handleRetry = useCallback(
		(callback?: () => void) => {
			retryFunctionReference.current = callback;
			setShowRetry(!!callback);
		},
		[retryFunctionReference, setShowRetry],
	);

	const handleFinish = useCallback(() => {
		if (isMultiple) {
			history.push(`/profiles/${activeProfile.id()}/dashboard`);
			return;
		}

		const importedWallet = activeProfile
			.wallets()
			.findByAddressWithNetwork(importedWallets[0].address, getValues("network").id());

		assertWallet(importedWallet);
		history.push(`/profiles/${activeProfile.id()}/wallets/${importedWallet.id()}`);
	}, [isMultiple, history, activeProfile, getValues, importedWallets]);

	return (
		<Tabs activeId={activeTab}>
			<StepIndicator size={4} activeIndex={activeTab} />

			<div data-testid="LedgerTabs" className="mt-8">
				<TabPanel tabId={LedgerTabStep.NetworkStep}>
					<NetworkStep
						filter={(network) =>
							network.allows(Enums.FeatureFlag.TransactionTransferLedgerS) ||
							network.allows(Enums.FeatureFlag.TransactionTransferLedgerX)
						}
						profile={activeProfile}
						title={t("WALLETS.PAGE_IMPORT_WALLET.NETWORK_STEP.TITLE")}
						subtitle={t("WALLETS.PAGE_IMPORT_WALLET.NETWORK_STEP.SUBTITLE")}
					/>
				</TabPanel>
				<TabPanel tabId={LedgerTabStep.LedgerConnectionStep}>
					<LedgerConnectionStep onConnect={() => setActiveTab(LedgerTabStep.LedgerScanStep)} />
				</TabPanel>
				<TabPanel tabId={LedgerTabStep.LedgerScanStep}>
					<LedgerScanStep profile={activeProfile} setRetryFn={handleRetry} />
				</TabPanel>
				<TabPanel tabId={LedgerTabStep.LedgerImportStep}>
					<LedgerImportStep
						wallets={importedWallets}
						profile={activeProfile}
						onClickEditWalletName={onClickEditWalletName}
					/>
				</TabPanel>
			</div>

			<Paginator
				activeIndex={activeTab}
				isMultiple={isMultiple}
				isNextDisabled={isNextDisabled}
				isNextLoading={isSubmitting}
				onBack={handleBack}
				onFinish={handleFinish}
				onNext={handleNext}
				onRetry={retryFunctionReference.current}
				showRetry={showRetry}
				size={4}
			/>
		</Tabs>
	);
};
