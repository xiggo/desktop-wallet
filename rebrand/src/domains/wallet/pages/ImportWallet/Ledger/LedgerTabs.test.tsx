import Transport from "@ledgerhq/hw-transport";
import { Contracts } from "@payvo/sdk-profiles";
import userEvent from "@testing-library/user-event";
import nock from "nock";
import React, { useEffect } from "react";
import { FormProvider, useForm } from "react-hook-form";
import { Route } from "react-router-dom";

import { LedgerTabs } from "./LedgerTabs";
import { LedgerProvider, minVersionList } from "@/app/contexts";
import * as scanner from "@/app/contexts/Ledger/hooks/scanner.state";
import { env, getDefaultLedgerTransport, getDefaultProfileId, render, screen, waitFor } from "@/utils/testing-library";

const nextSelector = () => screen.getByTestId("Paginator__continue-button");
const backSelector = () => screen.getByTestId("Paginator__back-button");

describe("LedgerTabs", () => {
	let profile: Contracts.IProfile;
	let wallet: Contracts.IReadWriteWallet;
	let transport: typeof Transport;
	let publicKeyPaths: Map<string, string>;
	let onClickEditWalletName: jest.Mock;
	let getVersionSpy: jest.SpyInstance;

	beforeAll(() => {
		publicKeyPaths = new Map<string, string>();

		nock("https://ark-test.payvo.com/api")
			.get("/wallets")
			.query((parameters) => !!parameters.address)
			.reply(200, {
				data: [
					{
						address: "DJpFwW39QnQvQRQJF2MCfAoKvsX4DJ28jq",
						balance: "2",
					},
					{
						address: "DSyG9hK9CE8eyfddUoEvsga4kNVQLdw2ve",
						balance: "3",
					},
				],
				meta: {},
			})
			.get("/wallets")
			.query((parameters) => !!parameters.address)
			.reply(200, {
				data: [],
				meta: {},
			})
			.get("/wallets")
			.query((parameters) => !!parameters.address)
			.reply(200, {
				data: [],
				meta: {},
			});
	});

	beforeAll(async () => {
		profile = env.profiles().findById(getDefaultProfileId());
		await env.profiles().restore(profile);
		await profile.sync();

		wallet = profile.wallets().first();

		getVersionSpy = jest
			.spyOn(wallet.coin().ledger(), "getVersion")
			.mockResolvedValue(minVersionList[wallet.network().coin()]);

		await wallet.synchroniser().identity();

		onClickEditWalletName = jest.fn();

		transport = getDefaultLedgerTransport();
		jest.spyOn(transport, "listen").mockImplementationOnce(() => ({ unsubscribe: jest.fn() }));

		publicKeyPaths = new Map([
			["m/44'/1'/0'/0/0", "027716e659220085e41389efc7cf6a05f7f7c659cf3db9126caabce6cda9156582"],
			["m/44'/1'/0'/0/1", "03d3fdad9c5b25bf8880e6b519eb3611a5c0b31adebc8455f0e096175b28321aff"],
			["m/44'/1'/0'/0/2", "025f81956d5826bad7d30daed2b5c8c98e72046c1ec8323da336445476183fb7ca"],
			["m/44'/1'/0'/0/3", "024d5eacc5e05e1b05c476b367b7d072857826d9b271e07d3a3327224db8892a21"],
			["m/44'/1'/0'/0/4", "025d7298a7a472b1435e40df13491e98609b9b555bf3ef452b2afea27061d11235"],

			["m/44'/1'/1'/0/0", wallet.publicKey()!],
			["m/44'/1'/2'/0/0", "020aac4ec02d47d306b394b79d3351c56c1253cd67fe2c1a38ceba59b896d584d1"],
			["m/44'/1'/3'/0/0", "033a5474f68f92f254691e93c06a2f22efaf7d66b543a53efcece021819653a200"],
			["m/44'/1'/4'/0/0", "03d3c6889608074b44155ad2e6577c3368e27e6e129c457418eb3e5ed029544e8d"],
		]);

		jest.spyOn(wallet.coin(), "__construct").mockImplementation();
		jest.spyOn(wallet.coin().ledger(), "getExtendedPublicKey").mockResolvedValue(wallet.publicKey()!);
	});

	afterAll(() => {
		getVersionSpy.mockRestore();
	});

	const BaseComponent = ({ activeIndex }: { activeIndex: number }) => (
		<Route path="/profiles/:profileId">
			<LedgerProvider transport={transport}>
				<LedgerTabs activeIndex={activeIndex} onClickEditWalletName={onClickEditWalletName} />
			</LedgerProvider>
		</Route>
	);

	const Component = ({ activeIndex }: { activeIndex: number }) => {
		const form = useForm({
			defaultValues: {
				network: wallet.network(),
			},
			mode: "onChange",
		});

		const { register } = form;

		useEffect(() => {
			register("network");
		}, [register]);

		return (
			<FormProvider {...form}>
				<BaseComponent activeIndex={activeIndex} />
			</FormProvider>
		);
	};

	it("should render scan step", async () => {
		jest.useRealTimers();
		const getPublicKeySpy = jest
			.spyOn(wallet.coin().ledger(), "getPublicKey")
			.mockImplementation((path) => Promise.resolve(publicKeyPaths.get(path)!));

		render(<Component activeIndex={2} />, { routes: [`/profiles/${profile.id()}`] });

		expect(screen.getByTestId("LedgerConnectionStep")).toBeVisible();

		// Auto redirect to next step
		await expect(screen.findByTestId("LedgerScanStep")).resolves.toBeVisible();

		await waitFor(() => expect(screen.getAllByRole("row")).toHaveLength(6), { timeout: 3000 });

		await waitFor(() => expect(screen.getAllByRole("checkbox")).toHaveLength(2), { timeout: 4000 });

		getPublicKeySpy.mockReset();
	});

	it("should render connection step", async () => {
		const getPublicKeySpy = jest
			.spyOn(wallet.coin().ledger(), "getPublicKey")
			.mockResolvedValue(publicKeyPaths.values().next().value);

		let formReference: ReturnType<typeof useForm>;
		const Component = () => {
			const form = useForm({ mode: "onChange" });
			const { register } = form;

			useEffect(() => {
				register("network", { required: true });
			}, [register]);

			formReference = form;

			return (
				<FormProvider {...form}>
					<BaseComponent activeIndex={1} />
				</FormProvider>
			);
		};

		const { container, history } = render(<Component />, { routes: [`/profiles/${profile.id()}`] });

		await expect(screen.findByTestId("SelectNetwork")).resolves.toBeVisible();

		await waitFor(() => expect(nextSelector()).toBeDisabled());
		await waitFor(() => expect(backSelector()).toBeEnabled());

		formReference!.setValue("network", wallet.network(), { shouldDirty: true, shouldValidate: true });

		expect(container).toMatchSnapshot();

		const historySpy = jest.spyOn(history, "push").mockImplementation();

		userEvent.click(backSelector());

		expect(historySpy).toHaveBeenCalledWith(`/profiles/${profile.id()}/dashboard`);

		historySpy.mockRestore();

		await waitFor(() => expect(nextSelector()).toBeEnabled());

		userEvent.click(nextSelector());

		expect(screen.getByTestId("LedgerConnectionStep")).toBeVisible();

		// Auto redirect to next step
		await expect(screen.findByTestId("LedgerScanStep")).resolves.toBeVisible();

		userEvent.click(backSelector());

		await expect(screen.findByTestId("SelectNetwork")).resolves.toBeVisible();

		await waitFor(() => expect(nextSelector()).toBeEnabled());

		getPublicKeySpy.mockReset();
	});

	it("should redirect to first screen when clicking back", async () => {
		const getPublicKeySpy = jest
			.spyOn(wallet.coin().ledger(), "getPublicKey")
			.mockImplementation((path) => Promise.resolve(publicKeyPaths.get(path)!));

		render(<Component activeIndex={2} />, { routes: [`/profiles/${profile.id()}`] });

		expect(screen.getByTestId("LedgerConnectionStep")).toBeVisible();

		// Auto redirect to next step
		await expect(screen.findByTestId("LedgerScanStep")).resolves.toBeVisible();

		userEvent.click(backSelector());

		await expect(screen.findByTestId("SelectNetwork")).resolves.toBeVisible();

		await waitFor(() => expect(nextSelector()).toBeEnabled());

		getPublicKeySpy.mockReset();
	});

	it("should render finish step", async () => {
		const getPublicKeySpy = jest
			.spyOn(wallet.coin().ledger(), "getPublicKey")
			.mockImplementation((path) => Promise.resolve(publicKeyPaths.get(path)!));

		const { history } = render(<Component activeIndex={3} />, { routes: [`/profiles/${profile.id()}`] });

		await expect(screen.findByTestId("LedgerScanStep")).resolves.toBeVisible();

		await waitFor(() => expect(screen.getAllByRole("row")).toHaveLength(2), { timeout: 3000 });

		expect(profile.wallets().values()).toHaveLength(2);

		await waitFor(() => expect(screen.getAllByRole("checkbox")).toHaveLength(2));

		userEvent.click(nextSelector());

		await expect(screen.findByTestId("LedgerImportStep")).resolves.toBeVisible();

		// Import wallets before entering the last step
		expect(profile.wallets().values()).toHaveLength(3);

		// First address
		userEvent.click(screen.getAllByTestId("LedgerImportStep__edit-alias")[0]);

		expect(onClickEditWalletName).toHaveBeenCalledTimes(1);

		const historySpy = jest.spyOn(history, "push").mockImplementation();

		userEvent.click(screen.getByTestId("Paginator__finish-button"));

		expect(historySpy).toHaveBeenCalledWith(`/profiles/${profile.id()}/wallets/${profile.wallets().last().id()}`);

		historySpy.mockRestore();
		getPublicKeySpy.mockReset();
	});

	it("should render finish step multiple", async () => {
		const getPublicKeySpy = jest
			.spyOn(wallet.coin().ledger(), "getPublicKey")
			.mockImplementation((path) => Promise.resolve(publicKeyPaths.get(path)!));

		const scannerMock = jest.spyOn(scanner, "scannerReducer").mockReturnValue({
			selected: ["m/44'/1'/0'/0/0", "m/44'/1'/0'/0/1"],
			wallets: [
				{
					address: "DSxxu1wGEdUuyE5K9WuvVCEJp6zibBUoyt",
					path: "m/44'/1'/0'/0/0",
				},
				{
					address: "DQh2wmdM8GksEx48rFrXCtJKsXz3bM6L6o",
					path: "m/44'/1'/0'/0/1",
				},
			],
		});

		const { history } = render(<Component activeIndex={3} />, { routes: [`/profiles/${profile.id()}`] });

		await expect(screen.findByTestId("LedgerScanStep")).resolves.toBeVisible();

		await waitFor(() => expect(screen.getAllByRole("row")).toHaveLength(3), { timeout: 3000 });

		const walletsCountBefore = profile.wallets().count();

		await waitFor(() => expect(screen.getAllByRole("checkbox")).toHaveLength(3));

		expect(screen.getByTestId("Paginator__continue-button")).toBeEnabled();

		userEvent.keyboard("{enter}");

		await expect(screen.findByTestId("LedgerImportStep")).resolves.toBeVisible();

		expect(profile.wallets().count()).toBe(walletsCountBefore + 2);

		const historySpy = jest.spyOn(history, "push").mockImplementation();

		userEvent.keyboard("{enter}");

		expect(historySpy).toHaveBeenCalledWith(`/profiles/${profile.id()}/dashboard`);

		historySpy.mockRestore();
		getPublicKeySpy.mockReset();
		scannerMock.mockRestore();
	});

	it("should render scan step with failing fetch", async () => {
		jest.spyOn(wallet.ledger(), "scan").mockRejectedValue(new Error("Scan Failed"));

		const getPublicKeySpy = jest
			.spyOn(wallet.coin().ledger(), "getPublicKey")
			.mockImplementation((path) => Promise.resolve(publicKeyPaths.get(path)!));

		const { container } = render(<Component activeIndex={2} />, { routes: [`/profiles/${profile.id()}`] });

		expect(screen.getByTestId("LedgerConnectionStep")).toBeVisible();

		// Auto redirect to next step
		await expect(screen.findByTestId("LedgerScanStep")).resolves.toBeVisible();
		await expect(screen.findByTestId("LedgerScanStep__error")).resolves.toBeVisible();

		userEvent.click(screen.getByTestId("Paginator__retry-button"));

		await waitFor(() => expect(screen.queryAllByTestId("LedgerScanStep__amount-skeleton")).toHaveLength(0), {
			interval: 5,
		});

		await expect(screen.findByTestId("LedgerScanStep__error")).resolves.toBeVisible();

		expect(container).toMatchSnapshot();

		userEvent.click(backSelector());

		await expect(screen.findByTestId("SelectNetwork")).resolves.toBeVisible();

		getPublicKeySpy.mockReset();
	});
});
