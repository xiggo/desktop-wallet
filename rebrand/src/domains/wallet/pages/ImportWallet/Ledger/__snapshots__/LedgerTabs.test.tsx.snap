// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LedgerTabs should render connection step 1`] = `
<div>
  <div>
    <ul
      class="StepIndicator__StepWrapper-sc-1kbqtlh-1 cTsjIC"
    >
      <li
        class="StepIndicator__StepStyled-sc-1kbqtlh-0 bIEtwY"
      />
      <li
        class="StepIndicator__StepStyled-sc-1kbqtlh-0 dXZskg"
      />
      <li
        class="StepIndicator__StepStyled-sc-1kbqtlh-0 dXZskg"
      />
      <li
        class="StepIndicator__StepStyled-sc-1kbqtlh-0 dXZskg"
      />
    </ul>
    <div
      class="mt-8"
      data-testid="LedgerTabs"
    >
      <div
        data-testid="tab-pabel__active-panel"
      >
        <section
          data-testid="NetworkStep"
        >
          <div
            class="flex justify-between items-end bg-theme-background"
          >
            <div
              class="space-y-4"
            >
              <h1
                class="mb-0 text-2xl"
                data-testid="header__title"
              >
                Select a Cryptoasset
              </h1>
              <div
                class="text-theme-secondary-text"
                data-testid="header__subtitle"
              >
                Select a cryptoasset to import your existing wallet address.
              </div>
            </div>
          </div>
          <fieldset
            class="FormField__FormFieldStyled-sc-q4bblx-0 dFNJMk flex flex-col mt-8 space-y-2"
          >
            <label
              class="flex mb-2 text-sm font-semibold transition-colors duration-100 FormLabel text-theme-secondary-text"
              data-testid="FormLabel"
              for="network"
            >
              Cryptoasset
            </label>
            <div>
              <div
                aria-expanded="false"
                aria-haspopup="listbox"
                aria-owns="NetworkStep__network-menu"
                data-testid="SelectNetwork"
                role="combobox"
              >
                <label
                  for="NetworkStep__network-input"
                  id="NetworkStep__network-label"
                />
                
                <div
                  class="Input__InputWrapperStyled-sc-1i8sd1r-0 cLNVAR"
                >
                  <div
                    class="Circle__CircleWrapper-sc-d5u41q-0 bcplCx border-theme-secondary-200 dark:border-theme-secondary-700 text-theme-secondary-500 ring-theme-background"
                    data-testid="SelectNetworkInput__network"
                  />
                  <div
                    class="relative flex flex-1 h-full"
                  >
                    <input
                      aria-autocomplete="list"
                      aria-controls="NetworkStep__network-menu"
                      aria-labelledby="NetworkStep__network-label"
                      autocomplete="off"
                      class="Input__InputStyled-sc-1i8sd1r-1 hPntwa p-0 border-none bg-transparent focus:ring-0 no-ligatures w-full"
                      data-testid="SelectNetworkInput__input"
                      id="NetworkStep__network-input"
                      placeholder="Enter a cryptoasset name"
                      value=""
                    />
                    
                  </div>
                </div>
              </div>
              <div
                class=""
                data-testid="SelectNetwork__options"
              >
                <div
                  class="mt-6"
                >
                  <div
                    class="text-sm font-bold text-theme-secondary-400 dark:text-theme-secondary-700"
                  >
                    PUBLIC NETWORKS
                  </div>
                  <ul
                    aria-labelledby="NetworkStep__network-label"
                    class="grid grid-cols-7 gap-3 mt-3"
                    id="NetworkStep__network-menu"
                    role="listbox"
                  >
                    <li
                      class="inline-block cursor-pointer h-18"
                      data-testid="SelectNetwork__NetworkIcon--container"
                    >
                      <div
                        aria-label="ARK"
                        class="w-full h-full flex justify-center items-center border-2 rounded-xl border-theme-primary-100 dark:border-theme-secondary-800 text-theme-primary-600"
                        data-testid="NetworkIcon-ARK-ark.mainnet"
                      >
                        <div
                          class="sc-bdvvtL qGXhc"
                          data-testid="NetworkIcon__icon"
                          height="20"
                          width="20"
                        >
                          <svg>
                            ark.svg
                          </svg>
                        </div>
                      </div>
                    </li>
                  </ul>
                </div>
                <div
                  class="Divider-sc-17ehhxk-0 FRyyK border-theme-secondary-300 dark:border-theme-secondary-800"
                  type="horizontal"
                />
                <div
                  class="mt-6"
                >
                  <div
                    class="flex justify-between items-center"
                  >
                    <span
                      class="text-sm font-bold text-theme-secondary-400 dark:text-theme-secondary-700"
                    >
                      DEVELOPMENT NETWORKS
                    </span>
                    <span
                      data-testid="SelectNetwork__developmentNetworks"
                    >
                      <label
                        class="Toggle__Wrapper-sc-6s3wfu-0 efNWJo"
                      >
                        <input
                          class="Toggle__Input-sc-6s3wfu-1 WwplI"
                          data-testid="SelectNetwork__developmentNetworks-toggle"
                          type="checkbox"
                        />
                        <div
                          class="Toggle__Handle-sc-6s3wfu-2 cobHFC"
                        >
                          <span
                            class="Toggle__HandleInner-sc-6s3wfu-3 ddepvC"
                          />
                        </div>
                      </label>
                    </span>
                  </div>
                  <ul
                    aria-labelledby="NetworkStep__network-label"
                    class="grid grid-cols-7 gap-3 mt-3 hidden"
                    id="NetworkStep__network-menu"
                    role="listbox"
                  >
                    <li
                      class="inline-block cursor-pointer h-18"
                      data-testid="SelectNetwork__NetworkIcon--container"
                    >
                      <div
                        aria-label="ARK Devnet"
                        class="w-full h-full flex justify-center items-center border-2 rounded-xl border-theme-primary-100 dark:border-theme-secondary-800 text-theme-secondary-700"
                        data-testid="NetworkIcon-ARK-ark.devnet"
                      >
                        <div
                          class="sc-bdvvtL qGXhc"
                          data-testid="NetworkIcon__icon"
                          height="20"
                          width="20"
                        >
                          <svg>
                            ark.svg
                          </svg>
                        </div>
                      </div>
                    </li>
                  </ul>
                </div>
              </div>
            </div>
          </fieldset>
        </section>
      </div>
    </div>
    <div
      class="flex justify-between mt-10"
    >
      <div />
      <div
        class="flex space-x-3"
      >
        <button
          class="OriginalButton__StyledButton-sc-o0fs2j-0 cxgwmk"
          data-testid="Paginator__back-button"
          type="button"
        >
          <div
            class="flex items-center space-x-2"
          >
            Back
          </div>
        </button>
        <button
          class="OriginalButton__StyledButton-sc-o0fs2j-0 kkDhXe"
          data-testid="Paginator__continue-button"
          disabled=""
          type="button"
        >
          <div
            class="flex items-center space-x-2"
          >
            Continue
          </div>
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`LedgerTabs should render scan step with failing fetch 1`] = `
<div>
  <div>
    <ul
      class="StepIndicator__StepWrapper-sc-1kbqtlh-1 cTsjIC"
    >
      <li
        class="StepIndicator__StepStyled-sc-1kbqtlh-0 bIEtwY"
      />
      <li
        class="StepIndicator__StepStyled-sc-1kbqtlh-0 bIEtwY"
      />
      <li
        class="StepIndicator__StepStyled-sc-1kbqtlh-0 bIEtwY"
      />
      <li
        class="StepIndicator__StepStyled-sc-1kbqtlh-0 dXZskg"
      />
    </ul>
    <div
      class="mt-8"
      data-testid="LedgerTabs"
    >
      <div
        data-testid="tab-pabel__active-panel"
      >
        <section
          class="space-y-8"
          data-testid="LedgerScanStep"
        >
          <div
            class="flex justify-between items-end bg-theme-background"
          >
            <div
              class="space-y-4"
            >
              <h1
                class="mb-0 text-2xl"
                data-testid="header__title"
              >
                Accounts
              </h1>
              <div
                class="text-theme-secondary-text"
                data-testid="header__subtitle"
              >
                Select the addresses that you want to import.
              </div>
            </div>
          </div>
          <fieldset
            class="FormField__FormFieldStyled-sc-q4bblx-0 dFNJMk flex flex-col space-y-2"
          >
            <label
              class="flex mb-2 text-sm font-semibold transition-colors duration-100 FormLabel text-theme-secondary-text"
              data-testid="FormLabel"
              for="network"
            >
              Cryptoasset
            </label>
            <div>
              <div
                aria-expanded="false"
                aria-haspopup="listbox"
                aria-owns="ImportWallet__network-menu"
                data-testid="SelectNetwork"
                role="combobox"
              >
                <label
                  for="ImportWallet__network-input"
                  id="ImportWallet__network-label"
                />
                
                <div
                  class="Input__InputWrapperStyled-sc-1i8sd1r-0 ktoJbG"
                  disabled=""
                >
                  <div
                    aria-label="ARK Devnet"
                    class="Circle__CircleWrapper-sc-d5u41q-0 bcplCx border-theme-secondary-200 dark:border-theme-secondary-700 text-theme-secondary-500 ring-theme-background"
                    data-testid="SelectNetworkInput__network"
                  >
                    <div
                      class="sc-bdvvtL qGXhc"
                      data-testid="NetworkIcon__icon"
                      height="20"
                      width="20"
                    >
                      <svg>
                        ark.svg
                      </svg>
                    </div>
                  </div>
                  <div
                    class="relative flex flex-1 h-full"
                  >
                    <input
                      aria-autocomplete="list"
                      aria-controls="ImportWallet__network-menu"
                      aria-labelledby="ImportWallet__network-label"
                      autocomplete="off"
                      class="Input__InputStyled-sc-1i8sd1r-1 hPntwa p-0 border-none bg-transparent focus:ring-0 no-ligatures w-full text-theme-secondary-text"
                      data-testid="SelectNetworkInput__input"
                      disabled=""
                      id="ImportWallet__network-input"
                      placeholder="Enter a cryptoasset name"
                      value="ARK Devnet"
                    />
                    
                  </div>
                </div>
              </div>
              <div
                class=""
                data-testid="SelectNetwork__options"
              >
                <div
                  class="mt-6 hidden"
                >
                  <ul
                    aria-labelledby="ImportWallet__network-label"
                    class="grid grid-cols-7 gap-3 mt-3"
                    id="ImportWallet__network-menu"
                    role="listbox"
                  />
                </div>
                <div
                  class="mt-6 hidden"
                >
                  <ul
                    aria-labelledby="ImportWallet__network-label"
                    class="grid grid-cols-7 gap-3 mt-3"
                    id="ImportWallet__network-menu"
                    role="listbox"
                  />
                </div>
              </div>
            </div>
          </fieldset>
          <div
            class="flex overflow-hidden flex-col space-y-4 rounded-lg p-4 bg-theme-danger-50 dark:bg-theme-secondary-800 sm:space-y-0 sm:space-x-4 sm:flex-row sm:items-center"
          >
            <div
              class="h-11 w-11 flex flex-shrink-0 items-center justify-center rounded-lg text-white bg-theme-danger-400"
            >
              <div
                class="sc-bdvvtL qGXhc"
                height="20"
                width="20"
              >
                <svg>
                  circle-cross.svg
                </svg>
              </div>
            </div>
            <div
              class="flex flex-col w-full"
            >
              <div
                class="text-sm leading-relaxed break-words dark:text-theme-secondary-500"
              >
                <span
                  data-testid="LedgerScanStep__error"
                >
                  Scan Failed
                </span>
              </div>
            </div>
          </div>
        </section>
      </div>
    </div>
    <div
      class="flex justify-between mt-10"
    >
      <div>
        <button
          class="OriginalButton__StyledButton-sc-o0fs2j-0 cxgwmk"
          data-testid="Paginator__retry-button"
          type="button"
        >
          <div
            class="flex items-center space-x-2"
          >
            <div
              class="sc-bdvvtL fCrhkg"
              height="16"
              width="16"
            >
              <svg>
                arrow-rotate-right.svg
              </svg>
            </div>
            <span>
              Retry
            </span>
          </div>
        </button>
      </div>
      <div
        class="flex space-x-3"
      >
        <button
          class="OriginalButton__StyledButton-sc-o0fs2j-0 cxgwmk"
          data-testid="Paginator__back-button"
          type="button"
        >
          <div
            class="flex items-center space-x-2"
          >
            Back
          </div>
        </button>
        <button
          class="OriginalButton__StyledButton-sc-o0fs2j-0 kkDhXe"
          data-testid="Paginator__continue-button"
          disabled=""
          type="button"
        >
          <div
            class="flex items-center space-x-2"
          >
            Continue
          </div>
        </button>
      </div>
    </div>
  </div>
</div>
`;
